File: packages/family-domain/src/setup-lifecycle.ts
import {
  AuditRequirementState,
  DeviceTrustState,
  HouseholdMembershipState,
  HouseholdRole,
} from '@ocentra-parent/schema-domain/family-household-authority';
import { ParentActorRole } from '@ocentra-parent/schema-domain/family-reference-primitives';
import {
  type RecoveryAuthorizationInput,
  RecoveryAuthorizationInputSchema,
  RecoveryBundleState,
  RecoveryChildEvidenceAccessState,
  RecoveryDataCustodyHandoffState,
  type RecoveryDecision,
  RecoveryDecisionState,
  RecoveryDecisionSchema,
  RecoveryDeleteExportState,
  RecoveryFailureReason,
  RecoveryIdentityProofState,
  RecoveryKind,
  type RecoveryOperation,
  RecoveryOperationSchema,
  RecoveryState,
  RecoveryStateSchema,
  RecoverySupportChannel,
} from '@ocentra-parent/schema-domain/family-restore-lifecycle';
import {
  type SetupInvite,
  type SetupInviteAuthorizationInput,
  SetupInviteAuthorizationInputSchema,
  type SetupInviteDecision,
  SetupInviteDecisionSchema,
  SetupInviteFailureReason,
  SetupInvitePurpose,
  SetupInviteReplayState,
  SetupInviteSchema,
  SetupInviteState,
  type SetupInviteTargetRole,
  SetupInviteDecisionState,
  SetupRecoveryAbuseState,
  SetupRecoveryResponseTimingState,
} from '@ocentra-parent/schema-domain/family-setup-invite';

export function isSetupInviteActive(input: SetupInvite): boolean {
  const invite = SetupInviteSchema.parse(input);

  return invite.state === SetupInviteState.Pending;
}

export function doesSetupInviteMatchTargetRole(input: SetupInvite): boolean {
  const invite = SetupInviteSchema.parse(input);

  return purposeMatchesTargetRole(invite.purpose, invite.targetRole);
}

export function isSetupInviteSinglePurpose(input: SetupInvite): boolean {
  const invite = SetupInviteSchema.parse(input);

  return invite.singleUse && doesSetupInviteMatchTargetRole(invite);
}

export function authorizeSetupInvite(input: SetupInviteAuthorizationInput): SetupInviteDecision {
  const parsedInput = SetupInviteAuthorizationInputSchema.parse(input);

  if (parsedInput.inviteState !== SetupInviteState.Pending) {
    return rejectedSetupInvite(SetupInviteFailureReason.InviteNotActive);
  }

  if (!parsedInput.singleUse) {
    return rejectedSetupInvite(SetupInviteFailureReason.InviteNotSingleUse);
  }

  if (parsedInput.replayState !== SetupInviteReplayState.Fresh) {
    return rejectedSetupInvite(SetupInviteFailureReason.InviteReplayRejected);
  }

  if (parsedInput.abuseState === SetupRecoveryAbuseState.Throttled) {
    return rejectedSetupInvite(SetupInviteFailureReason.InviteNotActive);
  }

  if (parsedInput.responseTimingState !== SetupRecoveryResponseTimingState.Uniform) {
    return rejectedSetupInvite(SetupInviteFailureReason.InviteNotActive);
  }

  if (!parsedInput.sameFamily) {
    return rejectedSetupInvite(SetupInviteFailureReason.WrongHousehold);
  }

  if (!purposeMatchesTargetRole(parsedInput.purpose, parsedInput.targetRole)) {
    return rejectedSetupInvite(SetupInviteFailureReason.WrongTargetRole);
  }

  if (!inviterCanIssue(parsedInput.inviterRole, parsedInput.purpose)) {
    return rejectedSetupInvite(SetupInviteFailureReason.InviterNotAuthorized);
  }

  return SetupInviteDecisionSchema.parse({
    decisionState: SetupInviteDecisionState.Acceptable,
    auditRequirementState: AuditRequirementState.Required,
    failureReason: null,
  });
}

export function recoveryRequiresOwnerApproval(input: RecoveryOperation): boolean {
  const recovery = RecoveryOperationSchema.parse(input);

  return requiresOwnerApproval(recovery.kind, recovery.ownerApprovalRequired);
}

export function recoveryDataCustodyHandoffState(input: RecoveryOperation): RecoveryDataCustodyHandoffState {
  const recovery = RecoveryOperationSchema.parse(input);

  return dataCustodyHandoffState(
    recovery.kind,
    recovery.deleteExportHandoffRequired,
    recovery.deleteExportState
  );
}

export function recoveryRequiresAuditedSupport(input: RecoveryOperation): boolean {
  const recovery = RecoveryOperationSchema.parse(input);

  return recovery.supportChannel === RecoverySupportChannel.SupportAssisted;
}

export function recoveryCanAccessChildEvidence(input: RecoveryOperation): boolean {
  const recovery = RecoveryOperationSchema.parse(input);
  const hasHouseholdAuthority =
    recovery.requesterMembershipState === HouseholdMembershipState.Active &&
    (recovery.requestedBy.role === ParentActorRole.Parent || recovery.requestedBy.role === ParentActorRole.Guardian);

  return (
    hasHouseholdAuthority &&
    recovery.identityProofState === RecoveryIdentityProofState.Verified &&
    recovery.supportChannel !== RecoverySupportChannel.SupportAssisted
  );
}

export function deviceTrustStateForRecoveryState(state: RecoveryState): typeof DeviceTrustState[keyof typeof DeviceTrustState] {
  const parsedState = RecoveryStateSchema.parse(state);

  if (
    parsedState === RecoveryState.PendingIdentityProof ||
    parsedState === RecoveryState.OwnerApprovalRequired ||
    parsedState === RecoveryState.Approved
  ) {
    return DeviceTrustState.ResetRequired;
  }

  if (parsedState === RecoveryState.Completed) {
    return DeviceTrustState.Pending;
  }

  return DeviceTrustState.Revoked;
}

export function deviceTrustStateForRecoveryOperation(
  input: RecoveryOperation
): typeof DeviceTrustState[keyof typeof DeviceTrustState] {
  const recovery = RecoveryOperationSchema.parse(input);

  if (recovery.state === RecoveryState.Revoked) {
    return DeviceTrustState.Revoked;
  }

  if (recovery.bundleFailureReason !== null) {
    return DeviceTrustState.ResetRequired;
  }

  switch (recovery.bundleState) {
    case RecoveryBundleState.PreviewOnly:
    case RecoveryBundleState.ApplyPending:
    case RecoveryBundleState.PartialRestore:
    case RecoveryBundleState.Rejected:
    case RecoveryBundleState.ManualRequired:
      return DeviceTrustState.ResetRequired;
    case RecoveryBundleState.Applied:
      return dataCustodyHandoffState(
        recovery.kind,
        recovery.deleteExportHandoffRequired,
        recovery.deleteExportState
      ) === RecoveryDataCustodyHandoffState.None && recovery.state === RecoveryState.Completed
        ? DeviceTrustState.Pending
        : DeviceTrustState.ResetRequired;
    default:
      return deviceTrustStateForRecoveryState(recovery.state);
  }
}

export function evaluateRecoveryOperation(input: RecoveryAuthorizationInput): RecoveryDecision {
  const parsedInput = RecoveryAuthorizationInputSchema.parse(input);
  const ownerApprovalRequired = requiresOwnerApproval(parsedInput.kind, parsedInput.ownerApprovalRequired);
  const childEvidenceAccessState = recoveryChildEvidenceAccessState(
    parsedInput.requesterRole,
    parsedInput.sameFamily,
    parsedInput.supportChannel
  );
  const custodyHandoffState = dataCustodyHandoffState(
    parsedInput.kind,
    parsedInput.deleteExportHandoffRequired,
    RecoveryDeleteExportState.None
  );

  if (parsedInput.state === RecoveryState.Revoked) {
    return rejectedRecovery(
      RecoveryFailureReason.RecoveryNotActive,
      ownerApprovalRequired,
      childEvidenceAccessState,
      custodyHandoffState
    );
  }

  if (parsedInput.requesterRole !== HouseholdRole.SupportAdmin && !parsedInput.sameFamily) {
    return rejectedRecovery(
      RecoveryFailureReason.WrongHousehold,
      ownerApprovalRequired,
      childEvidenceAccessState,
      custodyHandoffState
    );
  }

  if (parsedInput.identityProofState !== RecoveryIdentityProofState.Verified) {
    return rejectedRecovery(
      RecoveryFailureReason.IdentityProofRequired,
      ownerApprovalRequired,
      childEvidenceAccessState,
      custodyHandoffState
    );
  }

  if (parsedInput.abuseState === SetupRecoveryAbuseState.Throttled) {
    return rejectedRecovery(
      RecoveryFailureReason.IdentityProofRequired,
      ownerApprovalRequired,
      childEvidenceAccessState,
      custodyHandoffState
    );
  }

  if (parsedInput.responseTimingState !== SetupRecoveryResponseTimingState.Uniform) {
    return rejectedRecovery(
      RecoveryFailureReason.IdentityProofRequired,
      ownerApprovalRequired,
      childEvidenceAccessState,
      custodyHandoffState
    );
  }

  if (!requesterCanRecover(parsedInput.requesterRole, parsedInput.kind, parsedInput.supportChannel)) {
    return rejectedRecovery(
      RecoveryFailureReason.RoleNotAuthorized,
      ownerApprovalRequired,
      childEvidenceAccessState,
      custodyHandoffState
    );
  }

  return RecoveryDecisionSchema.parse({
    decisionState: RecoveryDecisionState.Authorized,
    ownerApprovalRequired,
    auditRequirementState: AuditRequirementState.Required,
    childEvidenceAccessState,
    dataCustodyHandoffState: custodyHandoffState,
    failureReason: null,
  });
}

function rejectedSetupInvite(failureReason: SetupInviteFailureReason): SetupInviteDecision {
  return SetupInviteDecisionSchema.parse({
    decisionState: SetupInviteDecisionState.Rejected,
    auditRequirementState: AuditRequirementState.Required,
    failureReason,
  });
}

function rejectedRecovery(
  failureReason: RecoveryFailureReason,
  ownerApprovalRequired: boolean,
  childEvidenceAccessState: RecoveryChildEvidenceAccessState,
  dataCustodyHandoffState: RecoveryDataCustodyHandoffState
): RecoveryDecision {
  return RecoveryDecisionSchema.parse({
    decisionState: RecoveryDecisionState.Rejected,
    ownerApprovalRequired,
    auditRequirementState: AuditRequirementState.Required,
    childEvidenceAccessState,
    dataCustodyHandoffState,
    failureReason,
  });
}

function purposeMatchesTargetRole(
  purpose: SetupInvitePurpose,
  targetRole: SetupInviteTargetRole | HouseholdRole
): boolean {
  return (
    (purpose === SetupInvitePurpose.CoParentInvite && targetRole === HouseholdRole.CoParentGuardian) ||
    (purpose === SetupInvitePurpose.ObserverInvite && targetRole === HouseholdRole.Observer) ||
    (purpose === SetupInvitePurpose.ChildDevicePairing && targetRole === HouseholdRole.ChildDeviceAgent) ||
    (purpose === SetupInvitePurpose.HouseholdTransfer && targetRole === HouseholdRole.ParentOwner)
  );
}

function inviterCanIssue(role: HouseholdRole, purpose: SetupInvitePurpose): boolean {
  switch (purpose) {
    case SetupInvitePurpose.CoParentInvite:
    case SetupInvitePurpose.ObserverInvite:
    case SetupInvitePurpose.ChildDevicePairing:
      return role === HouseholdRole.ParentOwner || role === HouseholdRole.CoParentGuardian;
    case SetupInvitePurpose.HouseholdTransfer:
      return role === HouseholdRole.ParentOwner;
    default:
      return false;
  }
}

function requesterCanRecover(
  role: HouseholdRole,
  kind: RecoveryKind,
  supportChannel: typeof RecoverySupportChannel[keyof typeof RecoverySupportChannel]
): boolean {
  if (role === HouseholdRole.SupportAdmin) {
    return supportChannel === RecoverySupportChannel.SupportAssisted;
  }

  if (kind === RecoveryKind.HouseholdTransfer) {
    return role === HouseholdRole.ParentOwner;
  }

  return role === HouseholdRole.ParentOwner || role === HouseholdRole.CoParentGuardian;
}

function requiresOwnerApproval(kind: RecoveryKind, ownerApprovalRequired: boolean): boolean {
  return (
    ownerApprovalRequired ||
    kind === RecoveryKind.LostParentDevice ||
    kind === RecoveryKind.CompromisedAccount ||
    kind === RecoveryKind.HouseholdTransfer
  );
}

function dataCustodyHandoffState(
  kind: RecoveryKind,
  deleteExportHandoffRequired: boolean,
  deleteExportState: RecoveryDeleteExportState
): RecoveryDataCustodyHandoffState {
  if (kind === RecoveryKind.HouseholdTransfer) {
    return RecoveryDataCustodyHandoffState.HouseholdTransferHandoffRequired;
  }

  if (deleteExportHandoffRequired && deleteExportState !== RecoveryDeleteExportState.DeleteConfirmed) {
    return RecoveryDataCustodyHandoffState.ExportDeleteHandoffRequired;
  }

  return RecoveryDataCustodyHandoffState.None;
}

function recoveryChildEvidenceAccessState(
  requesterRole: HouseholdRole,
  sameFamily: boolean,
  supportChannel: typeof RecoverySupportChannel[keyof typeof RecoverySupportChannel]
): RecoveryChildEvidenceAccessState {
  const hasHouseholdAuthority =
    sameFamily &&
    (requesterRole === HouseholdRole.ParentOwner || requesterRole === HouseholdRole.CoParentGuardian);

  if (hasHouseholdAuthority && supportChannel !== RecoverySupportChannel.SupportAssisted) {
    return RecoveryChildEvidenceAccessState.Allowed;
  }

  return RecoveryChildEvidenceAccessState.Blocked;
}

File: crates/family-identity-core/tests/unit/setup_lifecycle.rs
use ocentra_family_identity_core::family_identity::{DeviceTrustState, HouseholdRole};
use ocentra_family_identity_core::household_authority::AuditRequirementState;
use ocentra_family_identity_core::setup_lifecycle::{
    authorize_setup_invite, device_trust_state_for_recovery_operation,
    device_trust_state_for_recovery_state, evaluate_recovery_operation,
    RecoveryChildEvidenceAccessState, RecoveryDataCustodyHandoffState, RecoveryDecisionState,
    RecoveryFailureReason, RecoveryIdentityProofState, RecoveryKind, RecoveryOperation,
    RecoveryState, RecoverySupportChannel, SetupInviteDecisionState, SetupInviteFailureReason,
    SetupInviteInput, SetupInvitePurpose, SetupInviteReplayState, SetupInviteState,
    SetupInviteTargetRole, SetupRecoveryAbuseState, SetupRecoveryResponseTimingState,
};

fn parent_member_invite(
    purpose: SetupInvitePurpose,
    target_role: SetupInviteTargetRole,
) -> SetupInviteInput {
    SetupInviteInput {
        inviter_role: HouseholdRole::ParentOwner,
        same_family: true,
        purpose,
        target_role,
        invite_state: SetupInviteState::Pending,
        single_use: true,
        replay_state: SetupInviteReplayState::Fresh,
        abuse_state: SetupRecoveryAbuseState::WithinLimit,
        response_timing_state: SetupRecoveryResponseTimingState::Uniform,
    }
}

fn parent_member_recovery(kind: RecoveryKind) -> RecoveryOperation {
    RecoveryOperation {
        requester_role: HouseholdRole::ParentOwner,
        same_family: true,
        kind,
        state: RecoveryState::Approved,
        owner_approval_required: false,
        identity_proof_state: RecoveryIdentityProofState::Verified,
        support_channel: RecoverySupportChannel::SelfServe,
        delete_export_handoff_required: false,
        abuse_state: SetupRecoveryAbuseState::WithinLimit,
        response_timing_state: SetupRecoveryResponseTimingState::Uniform,
    }
}

#[test]
fn active_parent_can_issue_matching_coparent_invite() {
    let decision = authorize_setup_invite(parent_member_invite(
        SetupInvitePurpose::CoParentInvite,
        SetupInviteTargetRole::CoParentGuardian,
    ));

    assert_eq!(
        decision.decision_state,
        SetupInviteDecisionState::Acceptable
    );
    assert_eq!(
        decision.audit_requirement_state,
        AuditRequirementState::Required
    );
    assert_eq!(decision.failure_reason, None);
}

#[test]
fn revoked_or_expired_invites_are_rejected() {
    let revoked = authorize_setup_invite(SetupInviteInput {
        invite_state: SetupInviteState::Revoked,
        ..parent_member_invite(
            SetupInvitePurpose::ObserverInvite,
            SetupInviteTargetRole::Observer,
        )
    });

    assert_eq!(revoked.decision_state, SetupInviteDecisionState::Rejected);
    assert_eq!(
        revoked.failure_reason,
        Some(SetupInviteFailureReason::InviteNotActive)
    );

    let expired = authorize_setup_invite(SetupInviteInput {
        invite_state: SetupInviteState::Expired,
        ..parent_member_invite(
            SetupInvitePurpose::CoParentInvite,
            SetupInviteTargetRole::CoParentGuardian,
        )
    });

    assert_eq!(expired.decision_state, SetupInviteDecisionState::Rejected);
    assert_eq!(
        expired.failure_reason,
        Some(SetupInviteFailureReason::InviteNotActive)
    );
}

#[test]
fn replayed_or_non_single_use_invites_are_rejected() {
    let replayed = authorize_setup_invite(SetupInviteInput {
        replay_state: SetupInviteReplayState::ReplayDetected,
        ..parent_member_invite(
            SetupInvitePurpose::CoParentInvite,
            SetupInviteTargetRole::CoParentGuardian,
        )
    });

    assert_eq!(replayed.decision_state, SetupInviteDecisionState::Rejected);
    assert_eq!(
        replayed.failure_reason,
        Some(SetupInviteFailureReason::InviteReplayRejected)
    );

    let reusable = authorize_setup_invite(SetupInviteInput {
        single_use: false,
        ..parent_member_invite(
            SetupInvitePurpose::ObserverInvite,
            SetupInviteTargetRole::Observer,
        )
    });

    assert_eq!(reusable.decision_state, SetupInviteDecisionState::Rejected);
    assert_eq!(
        reusable.failure_reason,
        Some(SetupInviteFailureReason::InviteNotSingleUse)
    );
}

#[test]
fn observer_and_child_device_pairing_invites_are_accepted_in_scope() {
    let observer = authorize_setup_invite(parent_member_invite(
        SetupInvitePurpose::ObserverInvite,
        SetupInviteTargetRole::Observer,
    ));

    assert_eq!(
        observer.decision_state,
        SetupInviteDecisionState::Acceptable
    );
    assert_eq!(
        observer.audit_requirement_state,
        AuditRequirementState::Required
    );
    assert_eq!(observer.failure_reason, None);

    let child_device_pairing = authorize_setup_invite(parent_member_invite(
        SetupInvitePurpose::ChildDevicePairing,
        SetupInviteTargetRole::ChildDeviceAgent,
    ));

    assert_eq!(
        child_device_pairing.decision_state,
        SetupInviteDecisionState::Acceptable
    );
    assert_eq!(
        child_device_pairing.audit_requirement_state,
        AuditRequirementState::Required
    );
    assert_eq!(child_device_pairing.failure_reason, None);
}

#[test]
fn child_device_pairing_invite_rejects_wrong_target_role() {
    let decision = authorize_setup_invite(parent_member_invite(
        SetupInvitePurpose::ChildDevicePairing,
        SetupInviteTargetRole::Observer,
    ));

    assert_eq!(decision.decision_state, SetupInviteDecisionState::Rejected);
    assert_eq!(
        decision.failure_reason,
        Some(SetupInviteFailureReason::WrongTargetRole)
    );
}

#[test]
fn non_member_household_cannot_accept_household_transfer_invite() {
    let decision = authorize_setup_invite(SetupInviteInput {
        same_family: false,
        purpose: SetupInvitePurpose::HouseholdTransfer,
        target_role: SetupInviteTargetRole::ParentOwner,
        ..parent_member_invite(
            SetupInvitePurpose::HouseholdTransfer,
            SetupInviteTargetRole::ParentOwner,
        )
    });

    assert_eq!(decision.decision_state, SetupInviteDecisionState::Rejected);
    assert_eq!(
        decision.failure_reason,
        Some(SetupInviteFailureReason::WrongHousehold)
    );
}

#[test]
fn guardian_cannot_issue_household_transfer_invite() {
    let decision = authorize_setup_invite(SetupInviteInput {
        inviter_role: HouseholdRole::CoParentGuardian,
        purpose: SetupInvitePurpose::HouseholdTransfer,
        target_role: SetupInviteTargetRole::ParentOwner,
        ..parent_member_invite(
            SetupInvitePurpose::HouseholdTransfer,
            SetupInviteTargetRole::ParentOwner,
        )
    });

    assert_eq!(decision.decision_state, SetupInviteDecisionState::Rejected);
    assert_eq!(
        decision.failure_reason,
        Some(SetupInviteFailureReason::InviterNotAuthorized)
    );
}

#[test]
fn lost_parent_device_recovery_requires_owner_approval() {
    let lost_parent_device = evaluate_recovery_operation(RecoveryOperation {
        kind: RecoveryKind::LostParentDevice,
        state: RecoveryState::OwnerApprovalRequired,
        ..parent_member_recovery(RecoveryKind::LostParentDevice)
    });

    assert_eq!(
        lost_parent_device.decision_state,
        RecoveryDecisionState::Authorized
    );
    assert!(lost_parent_device.owner_approval_required);
    assert_eq!(
        lost_parent_device.audit_requirement_state,
        AuditRequirementState::Required
    );
    assert_eq!(
        lost_parent_device.child_evidence_access_state,
        RecoveryChildEvidenceAccessState::Allowed
    );
    assert_eq!(
        lost_parent_device.data_custody_handoff_state,
        RecoveryDataCustodyHandoffState::None
    );
    assert_eq!(lost_parent_device.failure_reason, None);
}

#[test]
fn child_reinstall_recovery_does_not_require_owner_approval() {
    let child_reinstall =
        evaluate_recovery_operation(parent_member_recovery(RecoveryKind::ChildReinstall));

    assert_eq!(
        child_reinstall.decision_state,
        RecoveryDecisionState::Authorized
    );
    assert!(!child_reinstall.owner_approval_required);
    assert_eq!(child_reinstall.failure_reason, None);
}

#[test]
fn delete_export_recovery_routes_to_data_custody() {
    let decision = evaluate_recovery_operation(RecoveryOperation {
        kind: RecoveryKind::ForgotLogin,
        delete_export_handoff_required: true,
        ..parent_member_recovery(RecoveryKind::ForgotLogin)
    });

    assert_eq!(decision.decision_state, RecoveryDecisionState::Authorized);
    assert_eq!(
        decision.data_custody_handoff_state,
        RecoveryDataCustodyHandoffState::ExportDeleteHandoffRequired
    );
    assert_eq!(decision.failure_reason, None);
}

#[test]
fn compromised_account_and_household_transfer_recovery_keep_owner_approval_parity() {
    let compromised_account =
        evaluate_recovery_operation(parent_member_recovery(RecoveryKind::CompromisedAccount));

    assert_eq!(
        compromised_account.decision_state,
        RecoveryDecisionState::Authorized
    );
    assert!(compromised_account.owner_approval_required);
    assert_eq!(
        compromised_account.audit_requirement_state,
        AuditRequirementState::Required
    );
    assert_eq!(
        compromised_account.child_evidence_access_state,
        RecoveryChildEvidenceAccessState::Allowed
    );
    assert_eq!(
        compromised_account.data_custody_handoff_state,
        RecoveryDataCustodyHandoffState::None
    );
    assert_eq!(compromised_account.failure_reason, None);

    let household_transfer =
        evaluate_recovery_operation(parent_member_recovery(RecoveryKind::HouseholdTransfer));

    assert_eq!(
        household_transfer.decision_state,
        RecoveryDecisionState::Authorized
    );
    assert!(household_transfer.owner_approval_required);
    assert_eq!(
        household_transfer.audit_requirement_state,
        AuditRequirementState::Required
    );
    assert_eq!(
        household_transfer.child_evidence_access_state,
        RecoveryChildEvidenceAccessState::Allowed
    );
    assert_eq!(
        household_transfer.data_custody_handoff_state,
        RecoveryDataCustodyHandoffState::HouseholdTransferHandoffRequired
    );
    assert_eq!(household_transfer.failure_reason, None);
}

#[test]
fn support_assisted_recovery_is_audited_and_cannot_access_child_evidence() {
    let decision = evaluate_recovery_operation(RecoveryOperation {
        requester_role: HouseholdRole::SupportAdmin,
        same_family: false,
        kind: RecoveryKind::CompromisedAccount,
        support_channel: RecoverySupportChannel::SupportAssisted,
        ..parent_member_recovery(RecoveryKind::CompromisedAccount)
    });

    assert_eq!(decision.decision_state, RecoveryDecisionState::Authorized);
    assert_eq!(
        decision.audit_requirement_state,
        AuditRequirementState::Required
    );
    assert_eq!(
        decision.child_evidence_access_state,
        RecoveryChildEvidenceAccessState::Blocked
    );
    assert_eq!(decision.failure_reason, None);
}

#[test]
fn observer_and_revoked_recovery_paths_are_rejected() {
    let observer = evaluate_recovery_operation(RecoveryOperation {
        requester_role: HouseholdRole::Observer,
        ..parent_member_recovery(RecoveryKind::ForgotLogin)
    });

    assert_eq!(observer.decision_state, RecoveryDecisionState::Rejected);
    assert_eq!(
        observer.failure_reason,
        Some(RecoveryFailureReason::RoleNotAuthorized)
    );

    let revoked = evaluate_recovery_operation(RecoveryOperation {
        state: RecoveryState::Revoked,
        ..parent_member_recovery(RecoveryKind::ForgotLogin)
    });

    assert_eq!(revoked.decision_state, RecoveryDecisionState::Rejected);
    assert_eq!(
        revoked.failure_reason,
        Some(RecoveryFailureReason::RecoveryNotActive)
    );
}

#[test]
fn household_transfer_without_identity_proof_is_rejected_and_keeps_custody_handoff() {
    let decision = evaluate_recovery_operation(RecoveryOperation {
        kind: RecoveryKind::HouseholdTransfer,
        identity_proof_state: RecoveryIdentityProofState::Pending,
        ..parent_member_recovery(RecoveryKind::HouseholdTransfer)
    });

    assert_eq!(decision.decision_state, RecoveryDecisionState::Rejected);
    assert!(decision.owner_approval_required);
    assert_eq!(
        decision.data_custody_handoff_state,
        RecoveryDataCustodyHandoffState::HouseholdTransferHandoffRequired
    );
    assert_eq!(
        decision.failure_reason,
        Some(RecoveryFailureReason::IdentityProofRequired)
    );
}

#[test]
fn throttled_or_variable_timing_invite_is_rejected_as_inactive() {
    let throttled = authorize_setup_invite(SetupInviteInput {
        abuse_state: SetupRecoveryAbuseState::Throttled,
        ..parent_member_invite(
            SetupInvitePurpose::ObserverInvite,
            SetupInviteTargetRole::Observer,
        )
    });
    assert_eq!(
        throttled.failure_reason,
        Some(SetupInviteFailureReason::InviteNotActive)
    );

    let variable_timing = authorize_setup_invite(SetupInviteInput {
        response_timing_state: SetupRecoveryResponseTimingState::Variable,
        ..parent_member_invite(
            SetupInvitePurpose::ObserverInvite,
            SetupInviteTargetRole::Observer,
        )
    });
    assert_eq!(
        variable_timing.failure_reason,
        Some(SetupInviteFailureReason::InviteNotActive)
    );
}

#[test]
fn throttled_or_variable_timing_recovery_requires_identity_proof() {
    let throttled = evaluate_recovery_operation(RecoveryOperation {
        abuse_state: SetupRecoveryAbuseState::Throttled,
        ..parent_member_recovery(RecoveryKind::ForgotLogin)
    });
    assert_eq!(
        throttled.failure_reason,
        Some(RecoveryFailureReason::IdentityProofRequired)
    );

    let variable_timing = evaluate_recovery_operation(RecoveryOperation {
        response_timing_state: SetupRecoveryResponseTimingState::Variable,
        ..parent_member_recovery(RecoveryKind::ForgotLogin)
    });
    assert_eq!(
        variable_timing.failure_reason,
        Some(RecoveryFailureReason::IdentityProofRequired)
    );
}

#[test]
fn recovery_device_trust_projection_matches_canonical_states() {
    assert_eq!(
        device_trust_state_for_recovery_state(RecoveryState::PendingIdentityProof),
        DeviceTrustState::ResetRequired
    );
    assert_eq!(
        device_trust_state_for_recovery_state(RecoveryState::OwnerApprovalRequired),
        DeviceTrustState::ResetRequired
    );
    assert_eq!(
        device_trust_state_for_recovery_state(RecoveryState::Approved),
        DeviceTrustState::ResetRequired
    );
    assert_eq!(
        device_trust_state_for_recovery_state(RecoveryState::Completed),
        DeviceTrustState::Pending
    );
    assert_eq!(
        device_trust_state_for_recovery_state(RecoveryState::Revoked),
        DeviceTrustState::Revoked
    );
}

#[test]
fn completed_delete_export_recovery_keeps_device_reset_required() {
    let trust_state = device_trust_state_for_recovery_operation(RecoveryOperation {
        state: RecoveryState::Completed,
        delete_export_handoff_required: true,
        ..parent_member_recovery(RecoveryKind::ForgotLogin)
    });

    assert_eq!(trust_state, DeviceTrustState::ResetRequired);
}
